src/nvim/insexpand.c

```diff
@@ -3599,11 +3598,21 @@ static void ins_compl_update_shown_match(void)
 }
 
 /// Delete the old text being completed.
-void ins_compl_delete(void)
+void ins_compl_delete(bool new_leader)
 {
+  // Avoid deleting text that will be reinserted when changing leader. This
+  // allows marks present on the original text to shrink/grow appropriately.
+  int i = 0;
+  if (new_leader) {
+    char *leader = ins_compl_leader();
+    while (leader[i] == compl_orig_text[i] && leader[i] != NUL) {
+      i++;
+    }
```

Comment:
I think this should compare by characters instead of bytes, because `ins_bytes()` inserts the string by characters.